[ANGLE] Support variable rasterization rate map on Metal backend

djg opened the pull request:
#### 9345e0330dffc377b721ad854304c35ac2ea09e2
<pre>
[ANGLE] Support variable rasterization rate map on Metal backend
<a href="https://bugs.webkit.org/show_bug.cgi?id=255971">https://bugs.webkit.org/show_bug.cgi?id=255971</a>
<a href="https://rdar.apple.com/problem/108542336">rdar://problem/108542336</a>

Reviewed by NOBODY (OOPS!).

EXPERIMENTAL extention to ANGLE allowing GL to rendering with Metal
rasterization rate maps.

* Source/ThirdParty/ANGLE/doc/ExtensionSupport.md:
* Source/ThirdParty/ANGLE/extensions/ANGLE_variable_rasterization_rate_metal.txt: Added.
* Source/ThirdParty/ANGLE/include/GLES2/gl2ext_angle.h:
* Source/ThirdParty/ANGLE/include/platform/autogen/FeaturesMtl_autogen.h:
* Source/ThirdParty/ANGLE/include/platform/mtl_features.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/Extension_files.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/GL_EGL_WGL_loader.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/GL_EGL_entry_points.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/GLenum_value_to_string_map.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/interpreter_utils.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/proc_table.json:
* Source/ThirdParty/ANGLE/scripts/generate_entry_points.py:
* Source/ThirdParty/ANGLE/scripts/gl_angle_ext.xml:
* Source/ThirdParty/ANGLE/scripts/registry_xml.py:
* Source/ThirdParty/ANGLE/src/common/entry_points_enum_autogen.cpp:
(angle::GetEntryPointName):
* Source/ThirdParty/ANGLE/src/common/entry_points_enum_autogen.h:
* Source/ThirdParty/ANGLE/src/common/frame_capture_utils_autogen.cpp:
(angle::WriteParamCaptureReplay):
(angle::ParamTypeToString):
* Source/ThirdParty/ANGLE/src/common/frame_capture_utils_autogen.h:
(angle::GLMTLRasterizationRateMapANGLE&gt;):
(angle::AccessParamValue):
(angle::SetParamVal&lt;ParamType::TGLMTLRasterizationRateMapANGLE&gt;):
(angle::InitParamValue):
* Source/ThirdParty/ANGLE/src/common/gl_enum_utils_autogen.cpp:
(gl::GLenumToString):
* Source/ThirdParty/ANGLE/src/libANGLE/Context.cpp:
(gl::Context::bindMetalRasterizationRateMap):
* Source/ThirdParty/ANGLE/src/libANGLE/Context_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/State.cpp:
(gl::PrivateState::PrivateState):
(gl::PrivateState::setVariableRasterizationRateEnabled):
(gl::PrivateState::setVariableRasterizationRateMap):
(gl::PrivateState::setEnableFeature):
(gl::PrivateState::getEnableFeature const):
(gl::State::getPointerv const):
* Source/ThirdParty/ANGLE/src/libANGLE/State.h:
(gl::PrivateState::isVariableRasterizationRateEnabled const):
(gl::PrivateState::getVariableRasterizationRateMap const):
* Source/ThirdParty/ANGLE/src/libANGLE/capture/capture_gles_ext_autogen.cpp:
(gl::CaptureBindMetalRasterizationRateMapANGLE):
* Source/ThirdParty/ANGLE/src/libANGLE/capture/capture_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/gles_extensions_autogen.cpp:
(gl::GetExtensionInfoMap):
* Source/ThirdParty/ANGLE/src/libANGLE/gles_extensions_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/ContextImpl.cpp:
(rx::ContextImpl::bindMetalRasterizationRateMap):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/ContextImpl.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/gl/StateManagerGL.cpp:
(rx::StateManagerGL::syncState):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ContextMtl.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ContextMtl.mm:
(rx::ContextMtl::onDestroy):
(rx::ContextMtl::updateExtendedState):
(rx::ContextMtl::bindMetalRasterizationRateMap):
(rx::ContextMtl::setupDrawImpl):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/DisplayMtl.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/DisplayMtl.mm:
(rx::DisplayMtl::initializeExtensions const):
(rx::DisplayMtl::initializeFeatures):
(rx::DisplayMtl::supportsVariableRasterizationRate const):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
(rx::mtl::RenderCommandEncoder::setScissorRect):
(rx::mtl::RenderCommandEncoder::setRasterizationRateMap):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_common.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_render_utils.mm:
(rx::mtl::ClearUtils::setupClearWithDraw):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/vulkan/ContextVk.cpp:
* Source/ThirdParty/ANGLE/src/libANGLE/validationES2.cpp:
(gl::err::ValidCap):
(gl::ValidateMaxShaderCompilerThreadsKHR):
(gl::ValidateBindMetalRasterizationRateMapANGLE):
* Source/ThirdParty/ANGLE/src/libANGLE/validationES32.cpp:
(gl::ValidateGetPointerv):
* Source/ThirdParty/ANGLE/src/libANGLE/validationESEXT_autogen.h:
* Source/ThirdParty/ANGLE/src/libGLESv2/entry_points_gles_ext_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/entry_points_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_no_capture_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_vulkan_secondaries_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_with_capture_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/opengl32_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/opengl32_with_wgl_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_egl_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_glx_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_wgl_autogen.cpp:
* Source/ThirdParty/ANGLE/src/tests/angle_end2end_tests.gni:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/VariableRasterizationRateTestMetal.mm: Added.
* Source/ThirdParty/ANGLE/util/autogen/angle_features_autogen.cpp:
* Source/ThirdParty/ANGLE/util/autogen/angle_features_autogen.h:
* Source/ThirdParty/ANGLE/util/capture/frame_capture_replay_autogen.cpp:
(angle::ReplayTraceFunctionCall):
* Source/ThirdParty/ANGLE/util/capture/trace_gles_loader_autogen.cpp:
* Source/ThirdParty/ANGLE/util/capture/trace_gles_loader_autogen.h:
* Source/ThirdParty/ANGLE/util/capture/trace_interpreter_autogen.cpp:
(angle::ParseCallCapture):
* Source/ThirdParty/ANGLE/util/gles_loader_autogen.cpp:
* Source/ThirdParty/ANGLE/util/gles_loader_autogen.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b689ee69bbf58c6f5f9b846ffe6ed6cd1e9c318b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25949 "Failed to checkout and rebase branch from PR 13190") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4560 "Failed to checkout and rebase branch from PR 13190") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27230 "Failed to checkout and rebase branch from PR 13190") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28049 "Failed to checkout and rebase branch from PR 13190") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23764 "Failed to checkout and rebase branch from PR 13190") 
| | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6322 "Failed to checkout and rebase branch from PR 13190") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1988 "Failed to checkout and rebase branch from PR 13190") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/28049 "Failed to checkout and rebase branch from PR 13190") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26201 "Failed to checkout and rebase branch from PR 13190") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/6322 "Failed to checkout and rebase branch from PR 13190") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/27230 "Failed to checkout and rebase branch from PR 13190") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28629 "Failed to checkout and rebase branch from PR 13190") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/6322 "Failed to checkout and rebase branch from PR 13190") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/27230 "Failed to checkout and rebase branch from PR 13190") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/28629 "Failed to checkout and rebase branch from PR 13190") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/6322 "Failed to checkout and rebase branch from PR 13190") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/27230 "Failed to checkout and rebase branch from PR 13190") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/28629 "Failed to checkout and rebase branch from PR 13190") | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3094 "Failed to checkout and rebase branch from PR 13190") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/51/builds/1988 "Failed to checkout and rebase branch from PR 13190") | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4484 "Failed to checkout and rebase branch from PR 13190") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/27230 "Failed to checkout and rebase branch from PR 13190") | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3551 "Failed to checkout and rebase branch from PR 13190") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3412 "Failed to checkout and rebase branch from PR 13190") | | | 
<!--EWS-Status-Bubble-End-->

djg committed:
Variable Rasterization Rate Map

Change-Id: Idde09afb3921a502a2f0ba52bbef4d6202a71f0f

kenrussell commented on Source/ThirdParty/ANGLE/extensions/ANGLE_variable_rasterization_rate_metal.txt:
```diff
@@ -0,0 +1,62 @@
+
+    Accepted by the <pname> parameter of GetPointerv:
+
+        METAL_RASTERIZATION_RATE_MAP_BINDING_ANGLE   0x93A9
```
These will cause collisions with `GL_ANGLE_request_extension`. Looking at the internal ANGLE enum allocation doc, you could use 0x96BC and 0x96BD for these.

djg replied:
Thanks, @kenrussell. Are you able to allocate those enums in your internal registry?

kenrussell replied:
Yes - have just done so.

kenrussell commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/DisplayMtl.mm:
```diff
@@ -1401,6 +1408,12 @@ bool IsMetalDisplayAvailable()
+
+bool DisplayMtl::supportsVariableRasterizationRate() const
+{
+    return [mMetalDevice supportsRasterizationRateMapWithLayerCount:1];
```
Upstream ANGLE would need a macOS version check around this call, since Chrome still ships on macOS 10.13 but this API was only introduced in macOS 10.15.4+.

kenrussell commented on Source/ThirdParty/ANGLE/extensions/ANGLE_variable_rasterization_rate_metal.txt:
```diff
@@ -0,0 +1,62 @@
+    /*
+     * GLmtlRasterizationRateMapANGLE is an opaque handle to an id<MTLRasterizationRateMap>
+     */
+    typedef void* GLmtlRasterizationRateMapANGLE;
```
This capitalization seems a bit odd - would GLMTLRasterizationRateMapANGLE work?

djg replied:
I chose to follow the convention used by `GLeglClientBufferEXT`, `GLeglImageOES` for
`GLmtlRasterizationRateMapANGLE`. I'm not particularly set on the name but was trying to be consistent.

kenrussell reviewed (commented):
Haven't reviewed in detail yet, but would you consider uploading this to ANGLE's code review tool? The ANGLE team will do a better review than me of the integration of this extension into the backend.

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/State.h:
```diff
@@ -625,6 +625,11 @@ class State : angle::NonCopyable
     void setPixelLocalStorageActivePlanes(GLsizei n);
     GLsizei getPixelLocalStorageActivePlanes() const { return mPixelLocalStorageActivePlanes; }
 
+    // GL_ANGLE_variable_rasterization_rate_metal
```
Two variants:
GL_ANGLE_variable_rasterization_rate_metal
GL_ANGLE_metal_variable_rasterization_rate

djg replied:
That's what I get for renaming the extension 5 times!

djg replied:
@kkinnunen-apple @kenrussell

> GL_ANGLE_variable_rasterization_rate_metal

I've standardized on this variant, since there could be a general `GL_ANGLE_variable_rasterization_rate` or platform specific `GL_ANGLE_variable_rasterization_rate_xxx` in the future.

Unless you think that the differences between platforms is too great for there to ever be a commonality? (As witnessed by nVidia and QCOM shading rate extensions)

kenrussell replied:
`GL_ANGLE_variable_rasterization_rate_metal` sounds fine.

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/ContextImpl.h:
```diff
@@ -283,6 +283,10 @@ class ContextImpl : public GLImplFactory
                                                           const gl::PixelLocalStoragePlane[],
                                                           const GLenum storeops[]);
 
+    // GL_ANGLE_metal_variable_rate_map
```
third variant here

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/DisplayMtl.mm:
```diff
@@ -1401,6 +1408,17 @@ bool IsMetalDisplayAvailable()
+
+bool DisplayMtl::supportsVariableRasterizationRate() const
+{
+    if (@available(ios 13.0, macOS 10.15.4, *))
```
I think you could mark up the maccatalyst and tvos versions too

kenrussell commented on Source/ThirdParty/ANGLE/extensions/ANGLE_variable_rasterization_rate_metal.txt:
```diff
@@ -0,0 +1,62 @@
+
+Version
+
+    Last Modified Date: March 30, 2023
```
Want to update this to today's date?

kenrussell commented on Source/ThirdParty/ANGLE/extensions/ANGLE_variable_rasterization_rate_metal.txt:
```diff
@@ -0,0 +1,62 @@
+
+New Procedures and Functions
+
+    void BindMetalRasterizationRateMapANGLE(GLMTLRasterizationRateMapANGLE map);
```
Though this extension's incomplete and being prototyped, would you like to mention anything about the semantics of this new binding point? Should this define ownership of the MTLRasterizationRateMap? MTLRasterizationRateMap needs to define multiple layer maps if performing layered rendering - which I assume would mean using multiple draw buffers in OpenGL ES semantics. Should this mention errors, for example that subsequent draw calls will fail with INVALID_OPERATION if the number of color attachments on the bound framebuffer doesn't match the number of layer maps in the rasterization rate map? Does passing nullptr restore the GL state to its original value?

kenrussell commented on Source/ThirdParty/ANGLE/extensions/ANGLE_variable_rasterization_rate_metal.txt:
```diff
@@ -0,0 +1,62 @@
+    <target> parameter of Enablei, Disablei, IsEnabledi, GetBooleani_v, GetIntegeri_v, GetInteger64i_v, and by the <pname> parameter
+    of GetBooleanv, GetIntegerv, GetInteger64v, GetFloatv, 
+
+        VARIABLE_RASTERIZATION_RATE_ANGLE            0x96BC
```
What happens if `GL_VARIABLE_RASTERIZATION_RATE_ANGLE` is enabled, but no rasterization rate map is bound?

djg replied:
In the current implementation, if the feature is enable but no map is bound the GL renders as normal and no error is generated.

kenrussell replied:
Thanks - sounds fine.

kenrussell commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ContextMtl.mm:
```diff
@@ -1646,6 +1650,19 @@ GLint GetOwnershipIdentity(const egl::AttributeMap &attribs)
+    id<MTLRasterizationRateMap> rateMap = (__bridge id<MTLRasterizationRateMap>)(map);
+    if (rateMap && rateMap.device != mContextDevice.get())
+    {
+        return angle::Result::Stop;
```
Does this cause an OpenGL ES error to be signaled to the application? If not, should it?

djg replied:
I've made a note that this needs a test.

kenrussell commented on Source/ThirdParty/ANGLE/src/libANGLE/State.cpp:
```diff
@@ -2446,6 +2453,20 @@ void State::setShadingRate(GLenum rate)
 
+void State::setVariableRasterizationRate(bool enabled)
+{
+    mVariableRasterizationRate = enabled;
```
Maybe only set the dirty bit state if `enabled != mVariableRasterizationRate`?

kenrussell commented on Source/ThirdParty/ANGLE/src/libANGLE/State.cpp:
```diff
@@ -2446,6 +2453,20 @@ void State::setShadingRate(GLenum rate)
+
+void State::setVariableRasterizationRateBinding(GLMTLRasterizationRateMapANGLE map)
+{
+    mVariableRasterizationRateBinding = map;
```
Same for `map != mVariableRasterizationRateBinding`?

kenrussell commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
```diff
@@ -1725,12 +1725,20 @@ inline void PopDebugGroupCmd(id<MTLRenderCommandEncoder> encoder, IntermediateCo
+        MTLCoordinate2DMake(mRenderPassMaxScissorRect.width, mRenderPassMaxScissorRect.height);
+    if (map)
+    {
+        maxScissorRect = [map mapPhysicalToScreenCoordinates:maxScissorRect forLayer:0];
```
Since forLayer is hardcoded to 0 here, does this mean this extension currently only works if the bound framebuffer only has one color attachment?

djg replied:
This is one area that is TBD and still in development.

kenrussell replied:
I see. Would you add a `TODO(djg):` here, and either point to an open issue in the extension spec, or describe the issue in detail?

kenrussell commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
```diff
@@ -1742,6 +1750,22 @@ inline void PopDebugGroupCmd(id<MTLRenderCommandEncoder> encoder, IntermediateCo
+    {
+        auto adjustedOrigin =
+            [map mapPhysicalToScreenCoordinates:MTLCoordinate2DMake(clampedRect.x, clampedRect.y)
+                                       forLayer:0];
```
Since forLayer's hardcoded to 0 here, too, it looks like this only works for framebuffers with a single color attachment.

kenrussell commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_render_utils.mm:
```diff
@@ -787,8 +787,9 @@ void SetupCommonBlitWithDrawStates(const gl::Context *context,
         GetViewport(params.dstRect, params.dstTextureSize.height, params.dstFlipY);
     MTLScissorRect scissorRectMtl =
         GetScissorRect(params.dstScissorRect, params.dstTextureSize.height, params.dstFlipY);
+    // TODO:
```
How big is this TODO to handle? Please reformat to `TODO(djg): ...` and describe more why the rasterization rate map isn't available here.

kenrussell commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_render_utils.mm:
```diff
@@ -1325,8 +1326,9 @@ ANGLE_INLINE void SetPipelineState(ComputeCommandEncoder *encoder,
 
     scissorRect = GetScissorRect(params.clearArea, params.dstTextureSize.height, params.flipY);
 
+    // TODO:
```
Same here.

kenrussell commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/vulkan/ContextVk.cpp:
```diff
@@ -5654,6 +5654,14 @@ angle::Result ContextVk::syncState(const gl::Context *context,
+                            // Nothing to do until ANGLE_variable_rasterization_rate_metal is
+                            // implemented.
+                            break;
+                        case gl::State::EXTENDED_DIRTY_BIT_VRR:
```
This case statement is duplicated. Surprised this compiled.

djg replied:
That wasn't intentional. Rebase issue?

kenrussell commented on Source/ThirdParty/ANGLE/src/tests/gl_tests/UniformBufferTest.cpp:
```diff
@@ -2801,6 +2801,78 @@ TEST_P(UniformBlockWithOneLargeArrayMemberTest, MemberTypeIsMatrix)
     checkResults(GLColor::green, GLColor::red, GLColor::green, GLColor::green);
 }
 
+#if 0
```
This looks like an `#ifdef`'d out copy of the previous test. Would you remove it?

djg replied:
This isn't supposed to be in this change. Reverting.

kenrussell commented on Source/ThirdParty/ANGLE/src/tests/gl_tests/VariableRasterizationRateTestMetal.mm:
```diff
@@ -0,0 +1,300 @@
+//
+// Copyright 2021 The ANGLE Project Authors. All rights reserved.
```
2023

djg replied:
🤦 That's what happens when copying an existing test to start from.

kenrussell commented on Source/ThirdParty/ANGLE/src/tests/gl_tests/VariableRasterizationRateTestMetal.mm:
```diff
@@ -0,0 +1,300 @@
+
+    ScopeMetalRasterizationRateMapRef(const ScopeMetalRasterizationRateMapRef &other)
+    {
+        if (mMap)
```
This is a copy constructor, so `this` is starting in a clean state - how would `mMap` be non-nil?

Did you mean to define `operator=`? If so and if you were aiming to properly manage refcounts, I think it's important to ref the incoming `other.mMap` before unref'ing `this`'s in order to handle self-assignment. Not 100% sure.

kenrussell commented on Source/ThirdParty/ANGLE/src/tests/gl_tests/VariableRasterizationRateTestMetal.mm:
```diff
@@ -0,0 +1,300 @@
+        return (__bridge id<MTLDevice>)reinterpret_cast<void *>(device);
+    }
+
+    template <size_t N, size_t M>
```
Might suggest X or W for the horizontal direction and Y or H for the vertical direction, just for ease of reading.

kenrussell reviewed (commented):
Nice work on this Dan. Mostly looks good but a few small comments; let me take a look again when they're resolved.

kenrussell commented on Source/ThirdParty/ANGLE/src/libANGLE/Context.cpp:
```diff
@@ -10218,6 +10218,12 @@ void Context::drawPixelLocalStorageEXTDisable(const PixelLocalStoragePlane plane
+void Context::bindMetalRasterizationRateMap(GLMTLRasterizationRateMapANGLE map)
+{
+    ANGLE_CONTEXT_TRY(mImplementation->bindMetalRasterizationRateMap(this, map));
+    mState.setVariableRasterizationRateBinding(map);
```
I don't know ANGLE's dirty bit mechanism well, but I'd somewhat assume that the binding should be a side effect of the dirty bits being flushed. Otherwise I'm not sure how the rasterization rate map gets set/reset between context switches.

kenrussell reviewed (approved):
Looks good to me as far as I can tell, but I'm sure I've missed things inside the ANGLE implementation. A few comments.

djg committed:
[ANGLE] Support variable rasterization rate map on Metal backend
https://bugs.webkit.org/show_bug.cgi?id=255971
rdar://problem/108542336

Reviewed by NOBODY (OOPS!).

EXPERIMENTAL extention to ANGLE allowing GL to rendering with Metal
rasterization rate maps.

* Source/ThirdParty/ANGLE/doc/ExtensionSupport.md:
* Source/ThirdParty/ANGLE/extensions/ANGLE_variable_rasterization_rate_metal.txt: Added.
* Source/ThirdParty/ANGLE/include/GLES2/gl2ext_angle.h:
* Source/ThirdParty/ANGLE/include/platform/autogen/FeaturesMtl_autogen.h:
* Source/ThirdParty/ANGLE/include/platform/mtl_features.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/Extension_files.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/GL_EGL_WGL_loader.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/GL_EGL_entry_points.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/GLenum_value_to_string_map.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/interpreter_utils.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/proc_table.json:
* Source/ThirdParty/ANGLE/scripts/generate_entry_points.py:
* Source/ThirdParty/ANGLE/scripts/gl_angle_ext.xml:
* Source/ThirdParty/ANGLE/scripts/registry_xml.py:
* Source/ThirdParty/ANGLE/src/common/entry_points_enum_autogen.cpp:
(angle::GetEntryPointName):
* Source/ThirdParty/ANGLE/src/common/entry_points_enum_autogen.h:
* Source/ThirdParty/ANGLE/src/common/frame_capture_utils_autogen.cpp:
(angle::WriteParamCaptureReplay):
(angle::ParamTypeToString):
* Source/ThirdParty/ANGLE/src/common/frame_capture_utils_autogen.h:
(angle::GLMTLRasterizationRateMapANGLE>):
(angle::AccessParamValue):
(angle::SetParamVal<ParamType::TGLMTLRasterizationRateMapANGLE>):
(angle::InitParamValue):
* Source/ThirdParty/ANGLE/src/common/gl_enum_utils_autogen.cpp:
(gl::GLenumToString):
* Source/ThirdParty/ANGLE/src/libANGLE/Context.cpp:
(gl::Context::bindMetalRasterizationRateMap):
* Source/ThirdParty/ANGLE/src/libANGLE/Context_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/State.cpp:
(gl::PrivateState::PrivateState):
(gl::PrivateState::setVariableRasterizationRateEnabled):
(gl::PrivateState::setVariableRasterizationRateMap):
(gl::PrivateState::setEnableFeature):
(gl::PrivateState::getEnableFeature const):
(gl::State::getPointerv const):
* Source/ThirdParty/ANGLE/src/libANGLE/State.h:
(gl::PrivateState::isVariableRasterizationRateEnabled const):
(gl::PrivateState::getVariableRasterizationRateMap const):
* Source/ThirdParty/ANGLE/src/libANGLE/capture/capture_gles_ext_autogen.cpp:
(gl::CaptureBindMetalRasterizationRateMapANGLE):
* Source/ThirdParty/ANGLE/src/libANGLE/capture/capture_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/gles_extensions_autogen.cpp:
(gl::GetExtensionInfoMap):
* Source/ThirdParty/ANGLE/src/libANGLE/gles_extensions_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/ContextImpl.cpp:
(rx::ContextImpl::bindMetalRasterizationRateMap):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/ContextImpl.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/gl/StateManagerGL.cpp:
(rx::StateManagerGL::syncState):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ContextMtl.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ContextMtl.mm:
(rx::ContextMtl::onDestroy):
(rx::ContextMtl::updateExtendedState):
(rx::ContextMtl::bindMetalRasterizationRateMap):
(rx::ContextMtl::setupDrawImpl):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/DisplayMtl.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/DisplayMtl.mm:
(rx::DisplayMtl::initializeExtensions const):
(rx::DisplayMtl::initializeFeatures):
(rx::DisplayMtl::supportsVariableRasterizationRate const):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
(rx::mtl::RenderCommandEncoder::setScissorRect):
(rx::mtl::RenderCommandEncoder::setRasterizationRateMap):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_common.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_render_utils.mm:
(rx::mtl::ClearUtils::setupClearWithDraw):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/vulkan/ContextVk.cpp:
* Source/ThirdParty/ANGLE/src/libANGLE/validationES2.cpp:
(gl::err::ValidCap):
(gl::ValidateMaxShaderCompilerThreadsKHR):
(gl::ValidateBindMetalRasterizationRateMapANGLE):
* Source/ThirdParty/ANGLE/src/libANGLE/validationES32.cpp:
(gl::ValidateGetPointerv):
* Source/ThirdParty/ANGLE/src/libANGLE/validationESEXT_autogen.h:
* Source/ThirdParty/ANGLE/src/libGLESv2/entry_points_gles_ext_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/entry_points_gles_ext_autogen.h:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_no_capture_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_vulkan_secondaries_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/libGLESv2_with_capture_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/opengl32_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/opengl32_with_wgl_autogen.def:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_egl_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_glx_autogen.cpp:
* Source/ThirdParty/ANGLE/src/libGLESv2/proc_table_wgl_autogen.cpp:
* Source/ThirdParty/ANGLE/src/tests/angle_end2end_tests.gni:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/VariableRasterizationRateTestMetal.mm: Added.
* Source/ThirdParty/ANGLE/util/autogen/angle_features_autogen.cpp:
* Source/ThirdParty/ANGLE/util/autogen/angle_features_autogen.h:
* Source/ThirdParty/ANGLE/util/capture/frame_capture_replay_autogen.cpp:
(angle::ReplayTraceFunctionCall):
* Source/ThirdParty/ANGLE/util/capture/trace_gles_loader_autogen.cpp:
* Source/ThirdParty/ANGLE/util/capture/trace_gles_loader_autogen.h:
* Source/ThirdParty/ANGLE/util/capture/trace_interpreter_autogen.cpp:
(angle::ParseCallCapture):
* Source/ThirdParty/ANGLE/util/gles_loader_autogen.cpp:
* Source/ThirdParty/ANGLE/util/gles_loader_autogen.h:
